diff --git a/dump/GraphDumper.cpp b/dump/GraphDumper.cpp
index 1a3539625..3905a31c7 100644
--- a/dump/GraphDumper.cpp
+++ b/dump/GraphDumper.cpp
@@ -7,6 +7,7 @@
 #include "CommitDumper.h"
 #include "GraphFileType.h"
 #include "FileUtils.h"
+#include "spdlog/spdlog.h"
 #include "versioning/Transaction.h"
 #include "versioning/VersionController.h"
 
@@ -17,6 +18,7 @@ namespace rv = rg::views;
 
 DumpResult<void> GraphDumper::dump(const Graph& graph, const fs::Path& path) {
     Profile profile {"GraphDumper::dump"};
+    spdlog::info("Dumping");
     if (path.exists()) {
         return DumpError::result(DumpErrorType::GRAPH_DIR_ALREADY_EXISTS);
     }
diff --git a/pipeline/operations/CreateGraphStep.cpp b/pipeline/operations/CreateGraphStep.cpp
index 99592fc94..3da6f4043 100644
--- a/pipeline/operations/CreateGraphStep.cpp
+++ b/pipeline/operations/CreateGraphStep.cpp
@@ -3,6 +3,7 @@
 #include <sstream>
 
 #include "ExecutionContext.h"
+#include "PipelineException.h"
 #include "Profiler.h"
 #include "SystemManager.h"
 
@@ -23,7 +24,9 @@ void CreateGraphStep::prepare(ExecutionContext* ctxt) {
 void CreateGraphStep::execute() {
     Profile profile {"CreateGraphStep::execute"};
 
-    _sysMan->createGraph(_graphName);
+    if(!_sysMan->createAndDumpGraph(_graphName)) {
+        throw PipelineException("Failed To Create Graph");
+    }
 }
 
 void CreateGraphStep::describe(std::string& descr) const {
diff --git a/storage/versioning/ChangeResult.cpp b/storage/versioning/ChangeResult.cpp
index 0b5659097..aba130537 100644
--- a/storage/versioning/ChangeResult.cpp
+++ b/storage/versioning/ChangeResult.cpp
@@ -6,7 +6,7 @@ std::string ChangeError::fmtMessage() const {
     if (_commitError) {
         return fmt::format("ChangeError: {} ({})",
                            ChangeErrorTypeDescription::value(_type),
-                           CommitErrorTypeDescription::value(_commitError->getType()));
+                           _commitError->fmtMessage());
     }
 
     return std::string {ChangeErrorTypeDescription::value(_type)};
diff --git a/storage/versioning/CommitResult.cpp b/storage/versioning/CommitResult.cpp
index 7ac0eddfa..34427034a 100644
--- a/storage/versioning/CommitResult.cpp
+++ b/storage/versioning/CommitResult.cpp
@@ -3,6 +3,12 @@
 namespace db {
 
 std::string CommitError::fmtMessage() const {
+    if(_dumpError) {
+        return fmt::format("Version controller error : {} ({})",
+                           CommitErrorTypeDescription::value(_type),
+                           _dumpError->fmtMessage());
+    }
+
     return fmt::format("Version controller error: {}",
                        CommitErrorTypeDescription::value(_type));
 }
diff --git a/storage/versioning/VersionController.cpp b/storage/versioning/VersionController.cpp
index 3f5bedabb..a27adcd41 100644
--- a/storage/versioning/VersionController.cpp
+++ b/storage/versioning/VersionController.cpp
@@ -3,6 +3,7 @@
 #include "Profiler.h"
 #include "Graph.h"
 #include "CommitView.h"
+#include "spdlog/spdlog.h"
 #include "versioning/Change.h"
 #include "versioning/CommitBuilder.h"
 #include "versioning/DataPartRebaser.h"
@@ -61,6 +62,7 @@ CommitResult<void> VersionController::submitChange(Change* change, JobSystem& jo
     Profile profile {"VersionController::submitChange"};
 
     std::scoped_lock lock(_mutex);
+    spdlog::info("we get here");
 
     const auto& head = _head.load();
     if (head->hash() != change->baseHash()) {
diff --git a/system/SystemManager.cpp b/system/SystemManager.cpp
index fac23b4ee..578f4285a 100644
--- a/system/SystemManager.cpp
+++ b/system/SystemManager.cpp
@@ -12,6 +12,7 @@
 #include "versioning/Transaction.h"
 #include "GMLImporter.h"
 #include "JobSystem.h"
+#include "DumpAndLoadManager.h"
 #include "GraphLoader.h"
 #include "FileUtils.h"
 #include "TuringConfig.h"
@@ -23,16 +24,68 @@ SystemManager::SystemManager(const TuringConfig& config)
     : _config(config),
     _changes(std::make_unique<ChangeManager>())
 {
-    _defaultGraph = createGraph("default");
+    init();
 }
 
 SystemManager::~SystemManager() {
 }
 
+void SystemManager::init() {
+    const auto list = _config.getGraphsDir().listDir();
+
+    if(std::find(list->begin(),list->end(), _config.getGraphsDir()/"default") != list->end()){
+        spdlog::info("loading default");
+        _defaultGraph = loadGraph("default");
+    }else {
+        spdlog::info("creating default");
+        _defaultGraph = createAndDumpGraph("default");
+    }
+
+    if(!_defaultGraph) {
+        throw TuringException("Could Not Initialise the Default Graph");
+    }
+}
+
+Graph* SystemManager::loadGraph(const std::string& name) {
+    std::unique_ptr<Graph> graph = Graph::create("default",_config.getGraphsDir()/"default");
+    auto* rawPtr = graph.get();
+
+    if(auto res = graph->getDumpAndLoadManager()->loadGraph(); !res) {
+        spdlog::info(res.error().fmtMessage());
+        return nullptr; 
+    }
+    if (!addGraph(std::move(graph), "default")) {
+        return nullptr;
+    }
+
+    return rawPtr;
+}
+
 Graph* SystemManager::createGraph(const std::string& name) {
     std::unique_ptr<Graph> graph = Graph::create(name,_config.getGraphsDir()/name);
     auto* rawPtr = graph.get();
 
+    if (auto res = graph->getDumpAndLoadManager()->dumpGraph(); !res) {
+        spdlog::info(res.error().fmtMessage());
+        return nullptr;
+    }
+
+    if (!addGraph(std::move(graph), name)) {
+        return nullptr;
+    }
+
+    return rawPtr;
+}
+
+Graph* SystemManager::createAndDumpGraph(const std::string& name) {
+    std::unique_ptr<Graph> graph = Graph::create(name,_config.getGraphsDir()/name);
+    auto* rawPtr = graph.get();
+
+    if (auto res = graph->getDumpAndLoadManager()->dumpGraph(); !res) {
+        spdlog::info(res.error().fmtMessage());
+        return nullptr;
+    }
+
     if (!addGraph(std::move(graph), name)) {
         return nullptr;
     }
@@ -40,6 +93,7 @@ Graph* SystemManager::createGraph(const std::string& name) {
     return rawPtr;
 }
 
+
 bool SystemManager::addGraph(std::unique_ptr<Graph> graph, const std::string& name) {
     std::unique_lock guard(_graphsLock);
 
diff --git a/system/SystemManager.h b/system/SystemManager.h
index ba0fb6a24..8140518b6 100644
--- a/system/SystemManager.h
+++ b/system/SystemManager.h
@@ -35,8 +35,14 @@ public:
 
     const TuringConfig& getConfig() const { return _config; }
 
+    void init();
+
+    Graph* loadGraph(const std::string& graphName);
+
     Graph* createGraph(const std::string& graphName);
 
+    Graph* createAndDumpGraph(const std::string& graphName);
+
     void listAvailableGraphs(std::vector<fs::Path>& names);
 
     void listGraphs(std::vector<std::string_view>& names);
